docs/rfc/007-endpoint-errors.md

```diff
@@ -0,0 +1,229 @@
+
+**Pros:**
+
+* Consumers now know the *expected* outcomes of calling an endpoint in a first-class way and can handle those outcomes accordingly.
+* Producers are able to express the different failure states of their endpoints in greater detail, and have conviction that those failure states are being handled by consumers.
+* Changing the failure states of an endpoint now visibly affects the API definition - changes in behavior are properly captured in code (rather than just documentation).
+
+**Cons:**
+
+* To be effective, these failure cases will need to be ‘forced’ upon consumers via the compiler. This could be as complex 
+  data types (perhaps unions of {success, error1, error2, ...} or via checked exceptions. This would entail some degree 
+  of (justified) work migrating to a with-errors world to wire each failure case into the correct behavior in client code.
+* Given the new flexibility, producers may more eagerly throw errors in their code, and ‘old’ clients will have a hard 
+  time dealing with these without upgrading.
+* Producers might be incentivized to cram *all* the possible outcomes of their endpoint as errors (see the 
+  `UserNotAuthorized` example above), and as such unduly burden consumers of their endpoint with things that should 
+  probably just be `RuntimeException`s. This in turn might have knock-on effects in the codebase depending on the exact
+  implementation used.
+* Introducing further types (whether as checked exceptions or otherwise) that determine control flow may result in them 
+  spreading throughout a code base and result in further coupling to the RPC layer. As an example, using checked exceptions
+  could encourage the consumer to throw the exception upwards rather than dealing with it at the callsite.
+  
+### Mandatory vs. Optional
+
+As discussed above, declared errors can be either made to be optional, or mandatory. Both versions come with their own
+pros and cons, and picking one side or the other inevitably involves making sacrifices.
+
+**The case for mandatory handling**
+
+_Forcing_ consumers to handle your exceptional cases is beneficial to the _producer_. You can have confidence that
+consumers will be encouraged to handle the exceptional cases, even if they are not reading your documentation. Acknowledging
+the presence of exceptional cases will be required at the minimum.
+
+There are benefits as a consumer as well - additional errors, behaviors, or edge cases will be exposed to you on upgrading
+your client, and you should (in theory) never miss out on a new case (assuming you keep your client relatively up to date).
+
+**The case for optional handling**
+
+Optional handling favors the _consumer_ - existing code and behaviors will not have to change at the outset, and they
+can opt into improved error handling as necessary. The obvious downside is that producers could publish new errors until
+the cows come home, but lazy consumers may never actually handle these edge cases.
```

Comment:
I think the downside is a little more subtle here. There is still value for producers defining more specific and rich errors since it helps clients better understand what happened. The cost is that clients will have a harder time _proactively_ handling error states